### PR TITLE
Ip end point source

### PR DESCRIPTION
**Pull request by AnthonySteele**
A simpler approach to resolving Ip+port endpoints
Moves some code out of `StatsDUdpTransport` so that it can be reused in other transports
And does not keep the DNS mapper around when it's not needed.

The outstanding question is - how was DNS caching working before? where did the `endpointCacheDuration` value come from, was the client supposed to set this up themselves if they wanted it? I don't think that this was simple before, but can it be simple now? What would be a reasonable value for a default DNS cache time?

### CODE REVIEW

File: src/JustEat.StatsD/EndpointLookups/CachedIpEndpointSource.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    /// </summary>
+    public class CachedIpEndpointSource : IPEndPointSource
+    {
+        private IPEndPoint _cachedValue;
+        private DateTime _expiry;
+        private readonly int _cacheDurationSeconds;
+
+        private readonly IPEndPointSource _inner;
+
+        public CachedIpEndpointSource(IPEndPointSource inner, int cacheDurationSeconds)
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/CachedIpEndpointSource.cs]: Personally I'd prefer a TimeSpan so calling code looks a bit more readable:

```csharp
new CachedIpEndpointSource(inner, 5 * 60);
```

vs.

```csharp
new CachedIpEndpointSource(inner, TimeSpan.FromMinutes(5));
```

It also allows the consumer arbitrary precision (though why you'd cache for less than one second I don't know).

File: src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        private readonly string _hostName;
+        private readonly int _port;
+
+        public DnsLookupIpEndpointSource(string hostName, int port)
+        {
+            _hostName = hostName;
+            _port = port;
+        }
+
+        public IPEndPoint Endpoint => new IPEndPoint(GetIpAddressOfHost(_hostName), _port);
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs]: This feels more like a method than a property to me as it does stuff that's non-trivial, plus if you accessed the value via the property more than once in the same method you'd get different values, which might be unexpected.

File: src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+        private IPAddress GetIpAddressOfHost(string hostName)
+        {
+            var endpoints = Dns.GetHostAddressesAsync(hostName)
+                .GetAwaiter().GetResult();
+            return endpoints[0];
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs]: Verify there's at least one so we don't get ```ArgumentOutOfRangeException```?

File: src/JustEat.StatsD/EndpointLookups/EndpointParser.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Net;
+
+namespace JustEat.StatsD.EndpointLookups
+{
+    public static class EndpointParser
+    {
+        public static IPEndPointSource MakeEndPointSource(string host, int port, int? endpointCacheDuration)
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/EndpointParser.cs]: Use ```TimeSpan?``` as per other comment?

File: src/JustEat.StatsD/EndpointLookups/EndpointParser.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            if (string.IsNullOrWhiteSpace(host))
+            {
+                throw new ArgumentException(nameof(host));
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/EndpointParser.cs]: Specify a message.

File: src/JustEat.StatsD/EndpointLookups/IPEndPointSource.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+﻿using System.Net;
+
+namespace JustEat.StatsD.EndpointLookups
+{
+    public interface IPEndPointSource
+    {
+        IPEndPoint Endpoint { get; }
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/IPEndPointSource.cs]: I think this should be a method as mentioned above.

File: src/JustEat.StatsD/StatsDUdpTransport.cs [hunk 7]
```diff
@@ -15,29 +14,16 @@ public class StatsDUdpTransport : IStatsDTransport
...
-        public StatsDUdpTransport(int endpointCacheDuration, string hostNameOrAddress, int port)
-            : this(new CachedDnsEndpointMapper(new DnsEndpointProvider(), endpointCacheDuration), hostNameOrAddress, port) {}
-
-        private StatsDUdpTransport(IDnsEndpointMapper endpointMapper, string hostNameOrAddress, int port)
+        public StatsDUdpTransport(IPEndPointSource endPointSource)
         {
-            _endPointMapper = endpointMapper;
-            _hostNameOrAddress = hostNameOrAddress;
-            _port = port;
+            _endpointSource = endPointSource;
```
Comment (martincostello) [inline src/JustEat.StatsD/StatsDUdpTransport.cs]: Null check.

File: src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs (hunk 2 above)
Comment (AnthonySteele) [inline src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs]: Fair enough, done.

File: src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs (hunk 3 above)
Comment (AnthonySteele) [inline src/JustEat.StatsD/EndpointLookups/DnsLookupIpEndpointSource.cs]: 👍

File: src/JustEat.StatsD/StatsDUdpTransport.cs (hunk 7 above)
Comment (AnthonySteele) [inline src/JustEat.StatsD/StatsDUdpTransport.cs]: 👍

File: src/JustEat.StatsD/EndpointLookups/CachedIpEndpointSource.cs (hunk 1 above)
Comment (AnthonySteele) [inline src/JustEat.StatsD/EndpointLookups/CachedIpEndpointSource.cs]: 👍

File: src/JustEat.StatsD/EndpointLookups/SimpleIpEndpoint.cs [hunk 8]
```diff
@@ -7,11 +7,16 @@ namespace JustEat.StatsD.EndpointLookups
...
+        private readonly IPEndPoint _value;
+
         public SimpleIpEndpoint(IPEndPoint value)
         {
-            Endpoint = value;
+            _value = value;
         }
 
-        public IPEndPoint Endpoint { get; }
+        public IPEndPoint GetEndpoint()
```
Comment (martincostello) [inline src/JustEat.StatsD/EndpointLookups/SimpleIpEndpoint.cs]: Minor stylistic point: maybe use ```public IPEndPoint GetEndpoint() => _value;``` ?

**Pull request merged**